Common/Logger.swift

```diff
@@ -50,98 +58,172 @@ class Logger {
+        if !fm.isWritableFile(atPath: currLog.path) {
             do {
-                try "".write(toFile: url.path, atomically: true, encoding: .utf8)
+                try "".write(toFile: currLog.path, atomically: true, encoding: .utf8)
             } catch {
-                zLog.error("Unable to create log file \(url.path)")
-                return false
+                zLog.error("Unable to create log file \(currLog.path)")
+                return
             }
         }
         
-        guard let lfh = FileHandle(forUpdatingAtPath: url.path) else {
-            zLog.error("Unable to get file handle for updating \(url.path)")
-            return false
+        // set redirects to go to to current log file
+        // do this here (rather than e.g., in init()) to handle the case where somebody deletes the log
+        // file. This allows logging to restart the next time this method is called)
+        guard let lfh = FileHandle(forUpdatingAtPath: currLog.path) else {
+            zLog.error("Unable to get file handle for updating \(currLog.path)")
+            return
         }
         
         lfh.seekToEndOfFile()
         let stdErrDes = dup2(lfh.fileDescriptor, FileHandle.standardError.fileDescriptor)
         let stdOutDes = dup2(lfh.fileDescriptor, FileHandle.standardOutput.fileDescriptor)
         
         if (stdErrDes != FileHandle.standardError.fileDescriptor) || (stdOutDes != FileHandle.standardOutput.fileDescriptor) {
-            zLog.error("Unable to capture stdout and stderr to file \(url.path)")
-            return false
+            zLog.error("Unable to capture stdout and stderr to file \(currLog.path)")
         }
         setbuf(__stdoutp, nil) // set stdout to flush
         
-        return true
+        // rotate logs (potentially))
+        if let attrs = try? fm.attributesOfItem(atPath: currLog.path) as NSDictionary {
+            let sz = attrs.fileSize()
+            
+            var dailyRotateNeeded = false
+            if rotateDaily, let lrt = lastRotateTime {
+                if !Calendar.current.isDate(lrt, inSameDayAs:Date()) {
+                    dailyRotateNeeded = true
+                }
+            }
+            
+            if force || dailyRotateNeeded || (sz >= Logger.FILE_SIZE_THRESHOLD) {
+                for i in (1...(Logger.MAX_NUM_LOGS-2)).reversed() {
+                    let from = currLog.appendingPathExtension("\(i)")
+                    let to = currLog.appendingPathExtension("\(i+1)")
+                    
+                    do {
+                        try? fm.removeItem(at: to)
+                        try fm.moveItem(at: from, to: to)
+                    } catch {
+                        zLog.error("Error rotating \(from.path) to \(to.path): \(error.localizedDescription)")
+                    }
+                }
+                
+                // roll out the current log
+                let to = currLog.appendingPathExtension("1")
+                do {
```

Comment:
shouldn't all this be guarded by a lock? seems like you need to pause logging for this moment?